# 1. Fasea: Problema formalizatzen

##Zertan datza komunitate detekzioa?
Komunitateen detekzioaren kontzeptua sareen zientzian sortu da, sistema konplexuen barruan taldeak aurkitzeko metodo gisa, grafiko batean irudikatuta. Deskonposizio-metodo tradizionalenek bloke diagonalen edo triangeluarren egitura zorrotza bilatzen dute, baina komunitateen detekzio-metodoek talde bereko nodoen lotura esanguratsua(handia) eta talde desberdinetako nodoen arteko lotura txikia duten azpisareak aurkitzen dituzte. Beraz, Community Detection problema sendoki elkar erlazionaturik dauden ertzen taldeak bilatzean datza talde desberdinetako ertzen loturak sendotasun txikia dutelarik. Erlazio hori ertzek dituzten propietateen bidez ezartzen da.

##Zertarako erabili daiteke komunitate detekzioa?

Gauza askotarako erabili daiteke, hala nola,  komunitateak detektatzeko teknikak baliagarriak dira sare sozialetako algoritmoek interes komunak dituzten pertsonak aurkitzeko eta estuki konektatuta edukitzeko. Beste adibide bat honako izan daiteke: komunitateen detekzioa ikaskuntza automatikoan erabiltzea antzeko propietateak dituzten taldeak detektatzeko eta hainbat arrazoirengatik taldeak ateratzeko.

##Komunitate bat definitzen

Komunitate kontzeptua hainbat alditan aipatu dugu, baina, nola defini genezake komunitate bat?

Esan genezake komunitate bat sare askoren propietate bat dela, zeinak horietako sare zehatz batek komunitate ugari izan ditzake eta komunitate baten barruko nodoak dentsitate handiz konektatuta egoten diren.

##Clustering vs Komunitate detekzioa

Komunitateen detekzioa clustering-aren antzekoa dela esan daiteke. Clusteringa ikaskuntza automatikoko teknika bat da, eta antzeko datu-puntuak kluster berean biltzen dira, bere atributuetan oinarrituta. Clusteringa sareei aplika dakiekeen arren, eremu zabalagoa da gainbegiratu gabeko ikaskuntza automatikoan, hainbat atributu motaz arduratzen dena. Bestalde, komunitateen detekzioa sareak aztertzeko bereziki diseinatuta dago, eta hori ertz izeneko atributu mota bakar baten mende dago. Gainera, clustering algoritmoek nodo periferiko bakarrak eta kide izan beharko luketen komunitateak bereizteko joera dute. Hala ere, clustering-teknikak eta komunitateak detektatzeko teknikak sareak aztertzeko arazo askori aplika dakizkieke, eta hainbat alde on eta txar plantea ditzakete eremuaren arabera.

##Komunitate detekziorako teknikak

Komunitateak detektatzeko metodoak, oro har, bi motatan sailka daitezke: metodo aglomeratiboak eta metodo zatitzaileak. Metodo aglomeratiboetan, ertzak banan-banan gehitzen zaizkio nodoak baino ez dituen grafiko bati. Ertzak indartsuenetik ahulenera gehitzen dira. Metodo zatitzaileak metodo aglomeratiboen aurkakoak dira. Horietan, ertzak banan-banan ezabatzen dira grafiko oso batetik.

Sare jakin batean edozein komunitate egon daiteke, eta komunitate horiek tamaina desberdinekoak izan daitezke. Ezaugarri horiek direla eta, komunitateak detektatzeko prozedura oso zaila da. Hala ere, hainbat teknika daude komunitateen detekzioaren arloan. Hona hemen lau proposamen oso popularrak direnak:

-Louvain Community Detection
-Surprise Community Detection
-Leiden Community Detection
-Walktrap Community Detection

##Nola adierazi behar ditugu komunitateak grafoan?

Guk soluzioa bektore batean errepresentatuko dugu, zeinak bektoreko i elementua grafoko nodo bat errepresentatuko duen eta hemen aurkitzen den balioa grafoko z komunitatea adieraziko du. 

##Bibliografia

https://www.sciencedirect.com/topics/computer-science/community-detection

https://towardsdatascience.com/community-detection-algorithms-9bd8951e7dae


#Datuen azterketa

Problema nola optimizatu pentsatzen hasi aurretik, komeni da eskura ditugun datuak ongi ulertzea eta ezagutzea. Horregatik, gure grafoaren ezaugarri garrantzitsuenak bistaratuko ditugu.

In [ ]:
import plotly.express as px
import sys
import numpy as np
import networkx as nx

In [ ]:
#pip3 install ipynb

In [ ]:
if __name__ == "__main__":
  
  from google.colab import drive
  drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
#directory = "/content/drive/MyDrive/Universidad/4.Maila/BH/BH laboak/BH Praktika/"
directory = "/content/drive/MyDrive/4.Kurtsoa/BH/BH Praktika/"
#directory = ""

sys.path.append(directory)

In [ ]:
from ipynb.fs.full.CDP_Sarrera_Ikasle import sortu_grafoa

In [ ]:
G = sortu_grafoa() #Grafoa sortu
n = G.number_of_nodes() #n= nodo kopurua
edges_num = G.number_of_edges()

In [ ]:
#Lortu nodoen graduak eta ertzen pisuak
if __name__ == "__main__":
  nodoak = list(G.nodes)
  graduak = dict()
  for i in range(len(nodoak)):
    gradua = G.degree(nodoak[i])
    if gradua not in graduak:
      graduak[gradua] = 1
    else:
      graduak[gradua] += 1

  edges = list(G.edges)

  pisuak = dict()
  for edge in edges:
    pisua = G[edge[0]][edge[1]]["weight"]
    if pisua not in pisuak:
      pisuak[pisua] = 1
    else:
      pisuak[pisua] += 1

In [ ]:
#Pantailaratu grafoari buruzko informazioa
if __name__ == "__main__":
  if 0 in graduak:
    isolatuak = graduak[0]
  else:
    isolatuak = 0

  degree_avg = sum([k*v for k, v in graduak.items()])/n
  weight_avg = sum([k*v for k, v in pisuak.items()])/n
  weigth_sum =  G.size(weight="weight")
  different_weights = len(pisuak)

  print("Nodo kopurua: ", n)
  print("Ertz kopurua: ", edges_num)
  print("Nodo bakoitzak duen graduen batazbestekoa: ", degree_avg)
  print("Ertzen pisuen batazbestekoa: ", weight_avg)
  print("Konexiorik gabeko nodoak: ", isolatuak)
  print("Grafoko ertz guztien pisuen batura: ", weigth_sum)
  print("Grafoko ertzen pisu desberdinen kopurua: ", different_weights)

Nodo kopurua:  1843
Ertz kopurua:  3215
Nodo bakoitzak duen graduen batazbestekoa:  3.4888768312533913
Ertzen pisuen batazbestekoa:  1.8979924036896365
Konexiorik gabeko nodoak:  0
Grafoko ertz guztien pisuen batura:  3498.0
Grafoko ertzen pisu desberdinen kopurua:  7


Ikusten den bezala, grafo instantzia honen nodo kopurua nahiko handia da. Soluzio posible guztiak onargarriak direla suposatuz, eta multzo konbinazio posible guztiak kontsideratuz, 
$k^{1843}$ soluzioko bilaketa espazio baten aurrean gaude, non $k=$ komunitate kopurua den. Hortaz, estrategia egokiak pentsatu beharko ditugu eta programa optimoak garatu. 

Horrez gain, nodoen graduen batazbestekoarekin ideia orokor bat egin dezakegu grafoaren konexutasunari buruz. Dirudienez, nodo askok ez dute loturarik elkarren artean. Baina ez dago nodo bakarturik ere, hau da, 0 gradua duen nodorik. Hauek berri onak dira Community Detection bezalako problema batentzat, nodo bakartuek komunitate kopuru handiagoak eska baititzakete. 

Azkenik, begietsi dezakegu grafoan ez daudela ertz asko nodo kopuruarekin kontrajarri ezkero. Haien pisuei buruz hasierako ideia bat ere egin dezakegu: 7 pisu balio ezberdin daude grafo osoan, eta batazbeste 1.9 ingurukoak dira. 

Gure grafoa oraindik hobeto ezagutzeko hainbat grafiko irudikatuko ditugu:

###Ertzen pisuen banaketa

In [ ]:
#Ertzen pisuen banaketaren barren grafika egin
if __name__ == "__main__":
  values = [v for v in pisuak.values()]
  keys = [k for k in pisuak.keys()]

  fig = px.bar( x= keys, y = values, labels=dict(x="Ertzaren pisua", y="Ertz kopurua"), log_y=True, title="Grafoko ertzen pisuen banaketa")
  fig.show()

  del(keys); del(values)

Ertzen pisuen batazbestekoak informazio gehiegi ematen ez duenez, pisuen banaketaren grafiko bat irudikatu dugu. Bertan begiesten da balioak [1,10] tartekoak direla, eta gehiengoak 1 pisukoak direla. Beraz, ez gara eskala handiko pisuekin arituko lanean.

###Nodoen graduen banaketa

In [ ]:
#Nodoen graduen banaketaren barren grafika egin
if __name__ == "__main__":
  values = [v for v in graduak.values()]
  keys = [k for k in graduak.keys()]

  fig = px.bar( x= keys, y = values, labels=dict(x="Nodoaren gradua", y="Nodo kopurua"), title="Grafoko nodoen graduen banaketa")
  fig.show()

  del(keys); del(values)

Kasu honetan balioen tartea altuagoa da, baina nodo gehienak 5 gradukoak edo gutxiagokoak dira. Hala ere, dirudienez badago nodoren bat 20 gradua baina gehiagokoa, baita 30 graduko bat ere. Community Detection probleman gradu altuko nodo horiek komunitate sendoak sortzeko abiapuntuak izan litezke. 

###Osagai konexuak

In [ ]:
if __name__ == "__main__":
  print("Grafo konexua da? ", nx.is_connected(G))
  print("Osagai konexu (azpigrafo konexu) kopurua: ", nx.number_connected_components(G))
  osagai_konexuak = nx.connected_components(G)

  osagai_tamainak = dict()
  for osagai in osagai_konexuak:
    tamaina = len(osagai)
    if tamaina not in osagai_tamainak:
      osagai_tamainak[tamaina] = 1
    else:
      osagai_tamainak[tamaina] +=1

  values = [v for v in osagai_tamainak.values()]
  keys = [str(k) for k in osagai_tamainak.keys()]
  keys.sort(key=int)

  fig = px.bar( x= keys, y = values, labels=dict(x="Osagai konexuaren nodo kopurua", y="Kopurua"), title="Grafoko osagai konexuen banaketa tamainaren arabera")
  fig.show()

  del(keys); del(values); del(osagai_tamainak); del(osagai_konexuak)

Grafo konexua da?  False
Osagai konexu (azpigrafo konexu) kopurua:  275


Dirudienez, ez gaude grafo konexu baten aurrean. Horrek esan nahi du elkarren artean konektatuta ez dauden azpigrafoz osatzen dela, zehazki 275. Teorian, horrek esan nahiko luke egokiena dela gutxienez komunitate kopurua 275ekoa izatea. Izan ere, elkarren artean konektatuta ez daude 2 azpigrafo multzo berean sartu ezkero, komunitate hori ez da sendoa izango. Dena den, komunitate kopuruaren egokitasunean ez gara gehiegi sartuko oraingoz.

Grafikoari begiratu ezkero, ikusi dezakegu kopuru dibertsitate nabariko azpigrafoak daudela, bereizgarriena 531 nodokoa izanik. Hori izango litzateke tamaina handieneko multzoa. 

#Soluzioen errepresentazioa

Behin gure problema ongi ezagututa, soluzioak errepresentatzeko kodeketa bat aukeratu behar dugu. Klasean hainbat kodeketa ikusi ditugu, horietako bat aproposa izan daiteke komunitateak identifikatzeko. Jarraian, proposatu soluzio batzuk esku artean dugun problemarentzako. <i>Nola kodetuko dugu komunitateen identifikazio bat?</i> Pentsatu...

In [ ]:
### Eman soluzioen adibide batzuk
soluzioa_1=[2, 0, 1, 1, 0, 0, 2, 1, 2, 2] #soluzioa_1[i] = grafoko i nodoa z multzokoa da. Kasu honetan grafoa 10 nodokoa da eta 3 komunitate daude.
soluzioa_2=[2, 0, 1, 1, 0, 0, 3, 1, 2, 2 , 3, 3, 1, 2, 2] #soluzioa_1[i] = grafoko i nodoa z multzokoa da. Kasu honetan grafoa 15 nodokoa da eta 4 komunitate daude.

# Helburu-funtzioa

Dagoeneko badakigu zer den komunitate bat, eta nola adierazi ere. Esan dugu, elkarlanen bitartez identifikatuko ditugula komunitateak. Baina, zerk eragiten du autore bat komunitate baten parte izateak? Adibidez, nahiko al da behin artikulu bat idaztea beste autore batzuekin, komunitatearen parte izateko? Erantzuna, argi dago, ezezkoa dela.

Guk gure helburu-funtzio propioa proposatzen dugu. Honen ideia komunitate bakoitzeko elkarrekin konektatuta dauden nodoen ertzen batura egitea da. Hau formula matematiko baten bidez horrela adieraziko genuke:

$$f(x)=\sum_{i=1}^n\sum_{j=i}^n 
\begin{cases} 
      w_{ij} & x_{i} = x_{j} \\
      0 & x_{i} \neq x_{j}
\end{cases}
$$

Horrela geratuko litzateke pseudokodean:

In [ ]:
#Pseudokodea
#def modularitateaCustom(G, soluzioa, weight='weight'):
    #value = 0
    # for nodo guztiak
        #for nodo guztiak aurrekotik hasita
          #Komunitate berdinekoak badira
            #value aldagaian gehitu bi nodoen arteko ertzaren pisua

    #return value

Eta honen inplementazioa:

In [ ]:
def modularitateaCustom(G, soluzioa, weight='weight'):
    value = 0
    ### Bete hemen 15-20 lerro
    for i in range(len(soluzioa)):
        for j in range(i,len(soluzioa)):
          if soluzioa[i]==soluzioa[j]:
            lag = G[nodoak[i]]
            if  nodoak[j] not in lag: #Ez dira auzokideak
              continue
            else:
              value+=lag[nodoak[j]][weight]
    return value

Honen arazo nagusia da denak komunitate berdin batean sartzeko joera duela, eta honen bidez helburu funtzioko oso balio altuak lortzen ditu. Beraz, honekin hurrengo bi puntuetan azaltzen dena lortzen dugu:
- Komunitate oso handi bat probabilitate askorekin oso kohesio txikia izateko. 
- Komunitate txiki asko kohesiorik gabekoak. 

Joera honen ondorioz, helburu funtzio hau baztertzea erabaki dugu. Horren ordez, problema berdinean aplikatu diren modularitate funtzio bat aukeratzea erabaki dugu.


Community Detection problemari buruzko informazioa bilatu ondoren, https://arxiv.org/abs/cond-mat/0408187 link-ean dagoen paperrean agertzen den helburu funtzioa erabiltzea erabaki dugu, hurrengo ekuazioak deskribatzen duena:

$$f(x)= \frac{1}{2m}\sum_{ij} \left[  A_{ij} - \frac{ k_ik_j } {2m}  \right] \delta(c_i,c_j)
$$

Helburu funtzio honen bidez komunitate baten trinkotasuna neurtu dezakegu, izan ere, komunitate bateko barruko ertzen batura handia izatea nahi dugu, beste komunitateekin dituen ertzen batura ahalik eta txikien mantenduz. Helburu funtztio honek hau egitea ahalbidetzen digu.

$f(x)$ modularitatea da, $\delta(c_i,c_j)$ funtzioak i nodoa j nodoaren auzokidea den esaten digu. 0 itzuliko du auzokidea ez bada bestela 1ekoa itzuliko du. $A_{ij}$ i nodotik j nodora dagoen ertzaren pisua adierazten digu. $k_i$ da i nodoa beste nodoekin konektatuta dagoen ertzen batura. $m$ grafoko ertz guztien batura adierazten digu, $m = \frac{1}{2} \sum_{ij}A_{ij}$ izanik.

Modularitatea sare bat zatitzea erabakitzen den moduaren propietate bat da: zatituta ez dauden sareek eta nodo bakoitza beren komunitatean jartzen dutenek zero modularitatea izango dute.

Gainera, irakurritako paper-ak honakoa esaten digu, erabilitako network-a ausazkoa baldin bada orduan modularitate funtzioak eman beharreko emaitza 0 izan beharko luke. Beraz, zero baino txikiagoak diren balioak emaitza txarrak izango dira. Zero baino handiagoak diren balioak ausazkotasunetik desbideratzeak (onera)adierazten dituzte, eta, praktikan, 0,3 inguruko balioa sare batean komunitate esanguratsu bat dagoela esan nahi du.

Hau da helburu funtzioaren pseudokodea:

In [ ]:
# def modularitateaOld(Grafoa, soluzioa)):
#   m = ertz guztien pisuen batura
#   loop soluzio osoa (i):
#     loop soluzio osoa (j):
#       Ez badira komunitate berekoak i eta j nodoak:
#         hurrengo kasura pasa
#       j nodoa ez bada i-ren auzokidea:
#         pisua =0
#       j nodoa i-ren auzokidea bada:
#         pisua = i eta j nodoa lotzen duen ertzaren pisua
#       degree_i = lortu i nodoaren gradua
#       degree_j = lortu j nodoaren gradua
#       value = pisua - degree_i*degree_j/(2*m)
#   itzuli value/(2*m)

Hau da helburu funtzioa inplementatuta:

In [ ]:
def modularitateaOld(G, soluzioa, weight='weight'):
    value = 0
    nodoak = list(G.nodes)
    m = G.size(weight="weight")/2
    for i in range(len(soluzioa)):
        for j in range(len(soluzioa)):
            if soluzioa[i] != soluzioa[j]:
                continue
                
            lag = G[nodoak[i]]
            if  nodoak[j] not in lag: #Ez dira auzokideak
                pisua = 0
            else:
                pisua = lag[nodoak[j]][weight]  
            
            degree_i = G.degree(nodoak[i]) 
            degree_j = G.degree(nodoak[j]) 
            value += pisua - degree_i*degree_j/(2*m)
    return value/(2*m)

Helburu funtzio hau oso motela da, eta horregatik bektorizatzea erabaki genuen. Ideia da pisuak eta $\frac{k_ik_j}{2m}$ balioak matrizeetan sartu eta np bidez hauen kenketa egitea. Modularitate funtzio bektorizatua UA taldearen eta Txintxarri eta kunplitu taldearen kolaborazioa izan da. Matrizeen hasieraketak horrela egiten ditugu: 

In [ ]:
if __name__ == "__main__":
  #Nodoen izenen lista (indizetik izenera pasatzeko)
  nodoak = list(G.nodes)

  #Grafoko ertz guztien pisuen batura
  m = G.size(weight="weight") 

  #Nodoen izenetatik indizera pasatzeko hiztegia
  indizeak = dict()
  ind = 0
  for nodo in nodoak:
    indizeak[nodo] = ind
    ind+=1

  #Lortu nodo batetik besterako ertzen pisuen matrizea (ertza ez badago pisua=0)
  pisuak = np.zeros((n,n))
  for i in range(n):
      for j in range(n):
        lag = G[nodoak[i]]
        if  nodoak[j] in lag: #dira auzokideak
            pisuak[i,j] = lag[nodoak[j]]["weight"]

  #Nodoen graduak eskuratzeko. Dena den, guk graduak erabili ordez nodoari konektatutako ertzen pisuen baturak erabiliko ditugu modularitatean.
  # graduak = []
  # for i in range(n):
  #   graduak.append(G.degree(nodoak[i]))

  #Nodo baten gradua beregana konektatutako ertz guztien pisuen batura kontsideratuko dugu
  graduak = []
  for i in range(n):
    graduak.append(np.sum(pisuak[i]))

  #Modularitate funtziorako aurrekalkulatu i eta j nodo pare posible guztien graduen biderketak 
  graduMul = np.zeros((n,n))
  for i in range(n):
      for j in range(n):
        graduMul[i,j] = graduak[i]*graduak[j]/(2*m)

Eta horrela geratuko litzateke gure modularitate funtzioa bektorizatuta:

In [ ]:
import numpy as np
def modularitatea(G, solution):
    n = G.number_of_nodes()
    ii, jj = np.ogrid[:n, :n]
    value = np.sum(np.where(solution[ii]==solution[jj], pisuak[ii,jj]-graduMul[ii,jj], 0)) 
    return value/(2*m)

In [ ]:
# pip install ipynb exekutatu liburutegi hau erabili ahal izateko.
from itertools import product
import sys
sys.path.append(directory)

import community
import networkx as nx

### Dei orokorrak
if __name__ == "__main__":
  
  G=sortu_grafoa()
  n = G.number_of_nodes()
  k = 10
  print("Nodo kopurua: ",n)

  ## Gure inplementazioa
  soluzioa_1 = np.random.randint(k,size=n)
  soluzioa_2 = np.random.randint(k,size=n)

  # partition1= dict(zip(G.nodes, soluzioa_1))
  # partition2= dict(zip(G.nodes, soluzioa_2))

  print("1. soluzioaren modularitatea: ", modularitatea(G, soluzioa_1))
  print("2. soluzioaren modularitatea: ", modularitatea(G, soluzioa_2))

Nodo kopurua:  1843
1. soluzioaren modularitatea:  -0.007341857684910861
2. soluzioaren modularitatea:  0.003912184075781597


In [ ]:
if __name__ == "__main__":
  partition1= dict(zip(G.nodes, soluzioa_1))
  print(partition1)
  print(G['Nisheeth Srivastava'])
  print(G.nodes)

{'Nan Li': 8, 'Rong Jin': 2, 'Zhi-Hua Zhou': 4, 'Fajwel Fogel': 8, "Alexandre d'Aspremont": 0, 'Milan Vojnovic': 0, 'Nisheeth Srivastava': 6, 'Ed Vul': 0, 'Paul R. Schrater': 1, 'Sewoong Oh': 9, 'Devavrat Shah': 7, 'Siu On Chan': 6, 'Ilias Diakonikolas': 3, 'Rocco A. Servedio': 9, 'Xiaorui Sun': 7, 'Aaron van den Oord': 9, 'Benjamin Schrauwen': 8, 'Robert A. Vandermeulen': 1, 'Clayton Scott': 5, 'Qiang Liu': 7, 'Alexander T. Ihler': 1, 'Joseph G. Makin': 0, 'Philip N. Sabes': 5, 'Isabel Valera': 8, 'Zoubin Ghahramani': 2, 'Kumar Dubey': 6, 'Qirong Ho': 0, 'Sinead A. Williamson': 1, 'Eric P. Xing': 9, 'Anqi Wu': 7, 'Mijung Park': 9, 'Oluwasanmi O. Koyejo': 5, 'Jonathan W. Pillow': 7, 'Balaji Lakshminarayanan': 6, 'Daniel M. Roy': 4, 'Yee Whye Teh': 0, 'Jason Chang': 3, 'John W. Fisher III': 4, 'Mehmet G?nen': 5, 'Adam A. Margolin': 8, 'Qichao Que': 6, 'Mikhail Belkin': 3, 'Yusu Wang': 7, 'Bo Dai': 3, 'Bo Xie': 4, 'Niao He': 7, 'Yingyu Liang': 2, 'Anant Raj': 4, 'Maria-Florina F. Balcan'